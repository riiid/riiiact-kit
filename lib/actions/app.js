export const TOGGLE_DRAWER = 'TOGGLE_DRAWER';

export const SHOW_SNACKBAR = 'SHOW_SNACKBAR';
export const HIDE_SNACKBAR = 'HIDE_SNACKBAR';

export const SHOW_PROGRESS = 'SHOW_PROGRESS';
export const HIDE_PROGRESS = 'HIDE_PROGRESS';

export const SHOW_TOOLBAR = 'SHOW_TOOLBAR';
export const HIDE_TOOLBAR = 'HIDE_TOOLBAR';

export const FIREBASE_AVAILABLE = 'FIREBASE_AVAILAVLE';

export const toggleDrawer = () => {
  return {
    type: TOGGLE_DRAWER
  };
};

export const showSnackbar = (message, props = {}) => {
  return {
    type: SHOW_SNACKBAR,
    message,
    props
  };
};

export const hideSnackbar = () => {
  return {
    type: HIDE_SNACKBAR
  };
};

export const showProgress = () => {
  return {
    type: SHOW_PROGRESS
  };
};

export const hideProgress = () => {
  return {
    type: HIDE_PROGRESS
  };
};

export const showToolbar = () => {
  return {
    type: SHOW_TOOLBAR
  };
};

export const hideToolbar = () => {
  return {
    type: HIDE_TOOLBAR
  };
};

export const firebaseAvailable = available => {
  return {
    type: FIREBASE_AVAILABLE,
    available
  };
};
