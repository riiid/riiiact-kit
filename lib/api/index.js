import initFirebase from 'api/firebase';

export const firebase = initFirebase();
