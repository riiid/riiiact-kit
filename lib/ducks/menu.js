export const REQUEST_MENU = 'REQUEST_MENU';
export const RECEIVE_MENU_SUCCESS = 'RECIEVE_MENU_SUCCESS';
export const RECEIVE_MENU_FAILURE = 'RECIEVE_MENU_FAILURE';

export const requestMenu = () => ({type: REQUEST_MENU});

export const receiveMenuSuccess = data => {
  return {
    type: RECEIVE_MENU_SUCCESS,
    data
  };
};

export const receiveMenuFailure = error => {
  return {
    type: RECEIVE_MENU_FAILURE,
    error
  };
};

const INITIAL_STATE = {
  count: {
    totalViews: 0,
    currentViews: 0
  }
};

export default (state = INITIAL_STATE, action) => {
  switch (action.type) {
    case RECEIVE_MENU_SUCCESS:
      return {
        ...state,
        count: action.data,
        error: null
      };
    case RECEIVE_MENU_FAILURE:
      return {
        ...state,
        count: {
          totalViews: 'N/A',
          currentViews: 'N/A'
        },
        error: action.error
      };
    default:
      return state;
  }
};
