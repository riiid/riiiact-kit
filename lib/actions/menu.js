import {firebase} from 'api';
import {showProgress, hideProgress} from 'actions/app';
import {keys} from 'lodash';
import {Observable} from 'rxjs/Observable';

export const RECEIVE_MENU_SUCCESS = 'RECIEVE_MENU_SUCCESS';
export const RECEIVE_MENU_FAILURE = 'RECIEVE_MENU_FAILURE';

export const receiveMenuSuccess = data => {
  return {
    type: RECEIVE_MENU_SUCCESS,
    data
  };
};

export const receiveMenuFailure = error => {
  return {
    type: RECEIVE_MENU_FAILURE,
    error
  };
};

const count = () => {
  const {db} = firebase;
  db.ref('stats').child('totalViews').transaction$(current => {
    if (current === null) {
      return 1;
    }
    return current + 1;
  }).subscribe();
};

export const subscribeMenu = () => {
  return dispatch => {
    const {db} = firebase;
    dispatch(showProgress());
    count();
    const tv$ = db.ref('stats').child('totalViews').observe$('value')
      .map(snapshot => snapshot.val());
    const cv$ = db.ref('online-users').observe$('value')
      .map(snapshot => snapshot.val());

    return Observable.combineLatest(tv$, cv$, (tv, cv) => {
      return {
        totalViews: tv,
        currentViews: keys(cv).length
      };
    })
    .subscribe(result => {
      dispatch(receiveMenuSuccess(result));
      dispatch(hideProgress());
    }, err => {
      dispatch(receiveMenuFailure(err));
      dispatch(hideProgress());
    });
  };
};
