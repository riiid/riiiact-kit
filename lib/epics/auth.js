import {Observable} from 'rxjs/Observable';
import {firebase} from 'api';
import {
  REQUEST_AUTH, receiveAuthSuccess, receiveAuthFailure
} from 'actions/auth';
import {showProgress, hideProgress, showSnackbar} from 'actions/app';

export const authEpic = (action$, store) => {
  const {auth, db} = firebase;
  const {dispatch} = store;
  return action$.ofType(REQUEST_AUTH)
    .do(() => dispatch(showProgress()))
    .mergeMap(() => {
      return auth.onAuthStateChanged$()
        .do(user => {
          if (user === null) {
            return auth.signInAnonymously();
          }
          dispatch(hideProgress());
          db.ref('online-users').child(user.uid).set(true);
          db.ref('online-users').child(user.uid).onDisconnect().set(null);
        })
        .map(receiveAuthSuccess)
        .catch(err => {
          dispatch(hideProgress());
          dispatch(showSnackbar('로그인 중 오류 발생'));
          return Observable.of(receiveAuthFailure(err));
        });
    });
};
