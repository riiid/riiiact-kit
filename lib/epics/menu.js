import {Observable} from 'rxjs/Observable';
import keys from 'lodash/keys';
import {firebase} from 'api';
import {
  REQUEST_MENU, receiveMenuSuccess, receiveMenuFailure
} from 'actions/menu';
import {showProgress, hideProgress} from 'actions/app';

const count = () => {
  const {db} = firebase;
  db.ref('stats').child('totalViews').transaction$(current => {
    if (current === null) {
      return 1;
    }
    return current + 1;
  }).subscribe();
};

export const menuEpic = (action$, store) => {
  const {db} = firebase;
  const {dispatch} = store;
  return action$.ofType(REQUEST_MENU)
    .do(() => dispatch(showProgress()))
    .mergeMap(() => {
      count();
      const tv$ = db.ref('stats').child('totalViews').observe$('value')
        .map(snapshot => snapshot.val());
      const cv$ = db.ref('online-users').observe$('value')
        .map(snapshot => snapshot.val());

      return Observable.combineLatest(tv$, cv$, (tv, cv) => {
        return {
          totalViews: tv,
          currentViews: keys(cv).length
        };
      })
      .do(() => dispatch(hideProgress()))
      .map(receiveMenuSuccess)
      .catch(err => {
        dispatch(hideProgress());
        return Observable.of(receiveMenuFailure(err));
      });
    });
};
