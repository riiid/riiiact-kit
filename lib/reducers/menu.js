import {RECEIVE_MENU_SUCCESS, RECEIVE_MENU_FAILURE} from 'actions/menu';

const INITIAL_STATE = {
  count: {
    totalViews: 0,
    currentViews: 0
  }
};

export default (state = INITIAL_STATE, action) => {
  switch (action.type) {
    case RECEIVE_MENU_SUCCESS:
      return {
        ...state,
        count: action.data,
        error: null
      };
    case RECEIVE_MENU_FAILURE:
      return {
        ...state,
        count: {
          totalViews: 'N/A',
          currentViews: 'N/A'
        },
        error: action.error
      };
    default:
      return state;
  }
};
