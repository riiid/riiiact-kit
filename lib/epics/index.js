import {combineEpics} from 'redux-observable';
import {authEpic} from 'epics/auth';
import {menuEpic} from 'epics/menu';

export default combineEpics(
  authEpic,
  menuEpic
);
