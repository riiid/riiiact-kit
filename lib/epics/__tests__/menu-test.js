jest.mock('api');
jest.mock('store');

import {Observable} from 'rxjs';
import createStore from 'store';
import {requestMenu} from 'actions/menu';
import {createEpicMiddleware} from 'redux-observable';
import {menu$$} from 'epics/menu';
import {firebase} from 'api';

describe('epics/menu', () => {
  let store;

  beforeEach(() => {
    store = createStore([
      createEpicMiddleware(menu$$, {dependencies: {firebase}})
    ])
  });


  it('when requestMenu success', () => {
    firebase.db.observe$.mockImplementationOnce(() => {
      return Observable.of({
        val() {
          return 100;
        }
      });
    });

    firebase.db.observe$.mockImplementationOnce(() => {
      return Observable.of({
        val() {
          return {key1: 'val1', key2: 'val2'};
        }
      });
    });

    firebase.db.transaction$.mockImplementationOnce(updateFn => {
      return Observable.of(updateFn(null));
    });

    store.dispatch(requestMenu());
    expect(store.getActions()).toMatchSnapshot();
  });

  it('when requestMenu fail', () => {
    firebase.db.transaction$.mockImplementationOnce(updateFn => {
      return Observable.of(updateFn({}));
    });

    firebase.db.observe$.mockImplementationOnce(() => {
      return Observable.throw(new Error('ERROR'));
    });

    firebase.db.observe$.mockImplementationOnce(() => {
      return Observable.throw(new Error('ERROR'));
    });

    store.dispatch(requestMenu());
    expect(store.getActions()).toMatchSnapshot();
  });
});
