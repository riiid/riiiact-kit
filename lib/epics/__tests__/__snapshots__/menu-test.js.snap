// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`epics/menu when requestMenu fail 1`] = `
Array [
  Object {
    "type": "REQUEST_MENU",
  },
  Object {
    "type": "SHOW_PROGRESS",
  },
  Object {
    "type": "HIDE_PROGRESS",
  },
  Object {
    "error": [Error: ERROR],
    "type": "RECIEVE_MENU_FAILURE",
  },
]
`;

exports[`epics/menu when requestMenu success 1`] = `
Array [
  Object {
    "type": "REQUEST_MENU",
  },
  Object {
    "type": "SHOW_PROGRESS",
  },
  Object {
    "type": "HIDE_PROGRESS",
  },
  Object {
    "data": Object {
      "currentViews": 2,
      "totalViews": 100,
    },
    "type": "RECIEVE_MENU_SUCCESS",
  },
]
`;
